### DISCUSSION TITLE
Doing more with the user id field

### DISCUSSION THREAD

**Discussion by aavetis** [Helicone/helicone#140] [Roadmap]
```
const response = await openai.createCompletion({
  model: "text-davinci-003",
  prompt: "How do I log users?",
  user: "alicebob@gmail.com",
});
```
Passing the `user` param is a great way to track metrics. Perhaps we can do more with the identified user?
- Rate limiting by user. See this [super simple implementation](https://github.com/Nutlope/roomGPT/blob/main/pages/api/generate.ts#L41-L46) of a user ID rate limiter implemented with Upstash
- Ability to block requests coming in from specific users. E.g. you identify malicious activity in their past requests.

**Comment by chitalian** [comment]
> Rate limiting by user. 

This is exactly what we want to add! What kind of toggles/knobs would you want to tune?

Maybe we can do requests/time cadence (ex: 1 request/ minute), or (100 requests / hour)

Like maybe the UX for this is like
```
`helicone-enable-user-rate-limit': 'true'
`helicone-user-requests`: "# of requests per cadence"
`helicone-user-cadence`: "# of seconds" 
```

Any other knobs around this, that you think you'd want to tune/turn? 

We can also make this configurable via the UI?

> Ability to block requests coming in from specific users. E.g. you identify malicious activity in their past requests.

This is cool! How do you expect this configuration to exist? I think the UI is the only clean way I can image this. We can build out the Users tab and add configuration settings there.

**Reply by aavetis** [reply]
> Any other knobs around this

Well those 2 params are pretty much [the only thing](https://github.com/Nutlope/roomGPT/blob/main/pages/api/generate.ts#L22-L29) people seem to use at the moment with redis caches :)

> This is cool! How do you expect this configuration to exist

Yeah, likely through the UI

**Comment by chitalian** [comment]
https://github.com/Helicone/helicone/issues/211

https://github.com/Helicone/helicone/issues/212
